kernel/arch/dreamcast/hardware/maple/controller.c

```diff
@@ -41,11 +48,54 @@ int cont_has_capabilities(const maple_device_t *cont, uint32_t capabilities) {
                    & capabilities) == capabilities) : -1;
 }
 
+static void * btn_callback_wrapper(void* args) {
+
+    for(;;) {
+        btn_callback(btn_callback_arg_addr, btn_callback_arg_btns);
+        thd_pass();
+    }
+
+    return NULL;
+}
+
+void cont_btn_callback_shutdown(void) {
+
+    /* This means either the callback is shutting down the 
+       whole system, or some jerk called this in the callback. */
+    if(thd_get_current()->tid == btn_callback_thd->tid) return;
+
+    thd_destroy(btn_callback_thd);
+    btn_callback_thd = NULL;
+    btn_callback = NULL;
+    btn_callback_addr = 0;
+    btn_callback_btns = 0;
+    btn_callback_arg_addr = 0;
+    btn_callback_arg_btns = 0;
+
+    return;
+}
+
 /* Set a controller callback for a button combo; set addr=0 for any controller */
 void cont_btn_callback(uint8_t addr, uint32_t btns, cont_btn_callback_t cb) {
+
+    /* Setting to NULL clears the current callback. */
+    if(cb == NULL) {
+        if(btn_callback_thd !=NULL)
+            cont_btn_callback_shutdown();
+        return;
+    }
+
     btn_callback_addr = addr;
     btn_callback_btns = btns;
     btn_callback = cb;
+
+    btn_callback_thd = thd_create(0, btn_callback_wrapper, NULL);
+
+    /* This may require an update to thd_create to be able to send custom flags. 
+       Otherwise it might run before it gets removed. */
+    thd_remove_from_runnable(btn_callback_thd);
```

Comment:
It's actually not safe to call this with interrupts enabled, as it could corrupt the runqueue if a thread scheduling interrupt comes in in the middle of this... Let me ponder the best way to deal with this (probably with another attribute in the `kthread_attr_t` struct).